redash/query_runner/yandex_metrika.py

```diff
@@ -0,0 +1,155 @@
+        if friendly_name in COLUMN_TYPES['date']:
+            data_type = TYPE_DATE
+        elif friendly_name in COLUMN_TYPES['datetime']:
+            data_type = TYPE_DATETIME
+        else:
+            data_type = TYPE_STRING
+        columns.append({'name': h, 'friendly_name': friendly_name, 'type': data_type})
+
+    rows = []
+    for num, row in enumerate(response['data']):
+        res = {}
+        for i, d in enumerate(row['dimensions']):
+            res[columns[i]['name']] = d['name']
+        for i, d in enumerate(row['metrics']):
+            res[columns[dimensions_len + i]['name']] = d
+            if num == 0 and isinstance(d, float):
+                columns[dimensions_len + i]['type'] = TYPE_FLOAT
+        rows.append(res)
+
+    return {'columns': columns, 'rows': rows}
+
+
+class YandexMetrika(BaseSQLQueryRunner):
+    @classmethod
+    def annotate_query(cls):
+        return False
+
+    @classmethod
+    def type(cls):
+        return "yandex_metrika"
+
+    @classmethod
+    def name(cls):
+        return "Yandex Metrika"
+
+    @classmethod
+    def configuration_schema(cls):
+        return {
+            "type": "object",
+            "properties": {
+                "token": {
+                    "type": "string",
+                    "title": "OAuth Token"
+                }
+            },
+            "required": ["token"],
+        }
+
+    def __init__(self, configuration):
+        super(YandexMetrika, self).__init__(configuration)
+        self.syntax = 'json'
+        self.host = 'https://api-metrika.yandex.ru'
+        self.list_path = 'counters'
+
+    def _get_tables(self, schema):
+
+        counters = self._send_query('management/v1/{0}'.format(self.list_path))
+
+        for row in counters[self.list_path]:
+            owner = row.get('owner_login')
+            counter = '{0} | {1}'.format(
+                row.get('name', 'Unknown').encode('utf-8'), row.get('id', 'Unknown')
+            )
+            if owner not in schema:
+                schema[owner] = {'name': owner, 'columns': []}
+
+            schema[owner]['columns'].append(counter)
+
+        return schema.values()
+
+    def test_connection(self):
+        self._send_query('management/v1/{0}'.format(self.list_path))
+
+    def _send_query(self, path='stat/v1/data', **kwargs):
+        token = kwargs.pop('oauth_token', self.configuration['token'])
+        r = requests.get('{0}/{1}'.format(self.host, path), params=dict(oauth_token=token, **kwargs))
+        if r.status_code != 200:
+            raise Exception(r.text)
+        return r.json()
+
+    def run_query(self, query, user):
+        logger.debug("Metrika is about to execute query: %s", query)
+        data = None
+        if query == "":
+            error = "Query is empty"
+            return data, error
+        try:
+            params = json.loads(query)
+        except ValueError:
+            params = parse_qs(urlparse(query).query, keep_blank_values=True)
```

Comment:
Maybe have a dedicated field for passing a URL, i.e.:

```json
{
    "url": "https://api-metrika.yandex.ru/stat/v1/data?dimensions=ym:s:searchEngine&date1=2016-01-01&date2=2018-03-28&metrics=ym:s:visits,ym:s:users&id=2138128&id=2138128&oauth_token=05dd3dd84ff948fdae2bc4fb91f13e22bb1f289ceef0037"
}
```

or if you decide to use YAML:

```yaml
url: https://api-metrika.yandex.ru/stat/v1/data?dimensions=ym:s:searchEngine&date1=2016-01-01&date2=2018-03-28&metrics=ym:s:visits,ym:s:users&id=2138128&id=2138128&oauth_token=05dd3dd84ff948fdae2bc4fb91f13e22bb1f289ceef0037
```

This way we won't have to guess if it's a URL/query string or something else.

Reply:
It's better to use string for our use case: make a loop over all counters and pass the same string with different "id" parameters.

I think it's better to keep this format. JSON/YAML for a separate queries because it looks good and url for execute_query method because it's compact.

Reply:
Unless I misunderstood you, the end result is the same, it just that you need to extract the URL from the JSON/YAML, so this section becomes:

```python
params = yaml.load(query)

if 'url' in params:
    params = parse_qs(urlparse(params['url']).query, keep_blank_values=True)
```

Reply:
I've talked about the further use of this query runner. 

Please look at our GA use case:

```
for acc in get_source_schema("Google Analytics"):
    view_id = 'ga:{0}'.format(int(app.split('(ga:')[1].split(')')[0]))
    res = execute_query(
        "Google Analytics",
        "https://www.googleapis.com/analytics/v3/data/ga?ids={0}&start-date={{start_date}}&end-date={{end_date}}&dimensions=ga:operatingSystem&metrics=ga:users,ga:sessions,ga:screenviews&samplingLevel=HIGHER_PRECISION&max-results=100".format(view_id)
    )
```

The string format looks better with python query runner's `execute_query()` method :) 

Of course I can add `url` key to parse it in query runner but I want to keep the ability to pass only string.

Reply:
Your example will change into:

```python
for acc in get_source_schema("Google Analytics"):
    view_id = 'ga:{0}'.format(int(app.split('(ga:')[1].split(')')[0]))
    res = execute_query(
        "Google Analytics",
        "url: https://www.googleapis.com/analytics/v3/data/ga?ids={0}&start-date={{start_date}}&end-date={{end_date}}&dimensions=ga:operatingSystem&metrics=ga:users,ga:sessions,ga:screenviews&samplingLevel=HIGHER_PRECISION&max-results=100".format(view_id)
    )
```

The reason I prefer only YAML/JSON format for the query is to avoid the case where we get a `ValueError` exception for some other reason (invalid format), try to parse query string (which we shouldn't) and then give cryptic error message to the user...